File: scala/src/main/scala/com/gildedrose/GildedRose.scala
package com.gildedrose

class GildedRose(val items: Array[Item]) {


  def updateQuality() {
    for (i <- 0 until items.length) {
      if (!items(i).name.equals("Aged Cheese")
        && !items(i).name.equals("Backstage passes to a concert")) {
        if (items(i).quality > 0) {
          if (!items(i).name.equals("Fine Italian Silk")) {
            items(i).quality = items(i).quality - 1
          }
        }
      } else {
        if (items(i).quality < 50) {
          items(i).quality = items(i).quality + 1

          if (items(i).name.equals("Backstage passes to a concert")) {
            if (items(i).sellIn < 11) {
              if (items(i).quality < 50) {
                items(i).quality = items(i).quality + 1
              }
            }

            if (items(i).sellIn < 6) {
              if (items(i).quality < 50) {
                items(i).quality = items(i).quality + 1
              }
            }
          }
        }
      }

      if (!items(i).name.equals("Fine Italian Silk")) {
        items(i).sellIn = items(i).sellIn - 1
      }

      if (items(i).sellIn < 0) {
        if (!items(i).name.equals("Aged Cheese")) {
          if (!items(i).name.equals("Backstage passes to a concert")) {
            if (items(i).quality > 0) {
              if (!items(i).name.equals("Fine Italian Silk")) {
                items(i).quality = items(i).quality - 1
              }
            }
          } else {
            items(i).quality = items(i).quality - items(i).quality
          }
        } else {
          if (items(i).quality < 50) {
            items(i).quality = items(i).quality + 1
          }
        }
      }
    }
  }
}

File: Kotlin/src/main/kotlin/com/gildedrose/GildedRose.kt
package com.gildedrose

class GildedRose(val items: List<Item>) {

    fun updateQuality() {
        for (i in items.indices) {
            if (items[i].name == "Sulfuras, Hand of Ragnaros") items[i].quality = 80;
            else {
                    if (items[i].name == "Aged Brie")
                        items[i].quality  = items[i].quality + 1
                    else {
                        if (items[i].name != "Backstage passes to a TAFKAL80ETC concert") {
                            items[i].quality = items[i].quality - 1
                            if (items[i].name.startsWith("Conjured")) items[i].quality -= 1
                        }
                        else {
                            items[i].quality = items[i].quality + 1

                            if (items[i].sellIn < 11)
                                items[i].quality = items[i].quality + 1

                            if (items[i].sellIn < 6)
                                items[i].quality = items[i].quality + 1
                        }
                    }

                    items[i].sellIn = items[i].sellIn - 1

                    if (items[i].sellIn < 0)
                    {
                        if (items[i].name == "Backstage passes to a TAFKAL80ETC concert")
                                        items[i].quality = 0
                        else
                            if (items[i].name == "Aged Brie")
                                items[i].quality = items[i].quality +1
                            else {
                                items[i].quality = items[i].quality - 1
                                if (items[i].name.startsWith("Conjured")) items[i].quality -= 1
                            }
                    }

                    if (items[i].quality > 50) items[i].quality = 50;
                    if (items[i].quality < 0) items[i].quality = 0;
                }
        }
    }
}


